modules/src/ics04_channel/handler/chan_open_try.rs

```diff
@@ -0,0 +1,479 @@
+use crate::ics04_channel::error::{Error, Kind};
+use crate::ics04_channel::handler::verify::verify_proofs;
+use crate::ics04_channel::handler::ChannelEvent::ChanOpenTry;
+use crate::ics04_channel::handler::ChannelResult;
+use crate::ics04_channel::msgs::chan_open_try::MsgChannelOpenTry;
+//use crate::ics24_host::identifier::ChannelId;
+
+pub(crate) fn process(
+    ctx: &dyn ChannelReader,
+    msg: MsgChannelOpenTry,
+) -> HandlerResult<ChannelResult, Error> {
+    let mut output = HandlerOutput::builder();
+
+    let channel_id;
+
+    // Unwrap the old channel end (if any) and validate it against the message.
+
+    let mut new_channel_end = match msg.previous_channel_id() {
+        Some(prev_id) => {
+            let old_channel_end = ctx
+                .channel_end(&(msg.port_id().clone(), prev_id.clone()))
+                .ok_or_else(|| Kind::ChannelNotFound.context(prev_id.to_string()))?;
+
+            channel_id = Some(prev_id.clone());
+            // Validate that existing channel end matches with the one we're trying to establish.
+
+            if old_channel_end.state_matches(&State::Init)
+                && old_channel_end.order_matches(&msg.channel.ordering())
+                && old_channel_end.connection_hops_matches(&msg.channel.connection_hops())
+                && old_channel_end.counterparty_matches(msg.channel.counterparty())
+               // && old_channel_end.version_matches(&msg.counterparty_version().clone())
+               && old_channel_end.version_matches(&msg.channel.version())
+            {
+                // A ChannelEnd already exists and all validation passed.
+                Ok(old_channel_end)
+            } else {
+                // A ConnectionEnd already exists and validation failed.
+                Err(Into::<Error>::into(
+                    Kind::ChannelMismatch(prev_id.clone()).context(
+                        old_channel_end
+                            .counterparty()
+                            .channel_id()
+                            .clone()
+                            .unwrap()
+                            .to_string(),
+                    ),
+                ))
+            }
+        }
+        // No channel id is supplied to  create a new channel end. Note: the id is assigned
+        // by the ChannelKeeper.
+        None => {
+            channel_id = None;
+
+            Ok(ChannelEnd::new(
+                State::Init,
+                *msg.channel.ordering(),
+                msg.channel.counterparty().clone(),
+                msg.channel.connection_hops().clone(),
+                msg.counterparty_version().clone(),
+            ))
+        }
+    }?;
+
+    // An IBC connection running on the local (host) chain should exist.
+
+    if msg.channel.connection_hops().len() != 1 {
+        return Err(Kind::InvalidConnectionHopsLength.into());
+    }
+
+    let connection_end = ctx.connection_state(&msg.channel().connection_hops()[0]);
```

Comment:
Can we change the method name to `connection_end` since it returns a `ConnectionEnd`?
```suggestion
    let connection_end = ctx.connection_end(&msg.channel().connection_hops()[0]);
```